core/state_prefetcher.go

```diff
@@ -59,9 +81,54 @@ func (p *statePrefetcher) Prefetch(block *types.Block, statedb *state.StateDB, c
 	)
 	workers.SetLimit(max(1, 4*runtime.NumCPU()/5)) // Aggressively run the prefetching
 
-	// Iterate over and process the individual transactions
-	for i, tx := range block.Transactions() {
+	var (
+		processed = make(map[common.Hash]struct{}, len(block.Transactions()))
+		heavyTxs  = make(chan *types.Transaction, len(block.Transactions()))
+		normalTxs = make(chan *types.Transaction, len(block.Transactions()))
+	)
+	for _, tx := range block.Transactions() {
+		// Note: gasLimit is not equivalent to gasUsed. Ideally, transaction heaviness
+		// should be measured using gasUsed. However, gasUsed is unknown prior to
+		// execution, so gasLimit is used as the indicator instead. This allows transaction
+		// senders to inflate gasLimit to gain higher prefetch priority, but this
+		// trade-off is unavoidable.
+		if isHeavyTransaction(tx.Gas(), block.GasUsed()) {
+			heavyTxs <- tx
+		}
+		// The heavy transaction will also be emitted with the normal prefetching
+		// ordering, depends on in which track it will be selected first.
+		normalTxs <- tx
+	}
+	blockPrefetchHeavyTxsMeter.Mark(int64(len(heavyTxs)))
+
+	fetchTx := func() (*types.Transaction, bool) {
+		// Pick the heavy transaction first based on the pre-defined probability
+		if rand.Intn(100) < heavyTransactionPriority {
+			select {
+			case tx := <-heavyTxs:
+				return tx, false
+			default:
+			}
+		}
+		// No more heavy transaction, or no priority for them, pick the transaction
+		// with normal order.
+		select {
+		case tx := <-normalTxs:
+			return tx, false
+		default:
+			return nil, true
+		}
+	}
+	for {
+		tx, done := fetchTx()
+		if done {
+			break
+		}
+		if _, exists := processed[tx.Hash()]; exists {
+			continue
+		}
 		stateCpy := statedb.Copy() // closure
+
 		workers.Go(func() error {
```

Comment:
There might be a concurrency issue with the `tx` variable here.  I think the intention here was for an updated `tx` value to be with each iteration. 

Iteration 1:
```
tx = 1
workers.Go(func() { use tx })
```

Iteration 2:
```
tx = 2
workers.Go(func() { use tx })
```

If the goroutine from iteration 1 doesn't run until after iteration 2, both goroutines may end up observing `tx == 2`. I think if you shadow `tx` variable before kicking off the goroutine, then it will fix it.  

```go
tx := tx
workers.Go(func() { use tx })

```